### ISSUE TITLE
Question: How to tokeninize text for Llama?

### ISSUE THREAD

**Issue by EricLBuehler** [huggingface/candle#944]
Hello everybody,

How can I tokenize text to use with Llama? I want to fine-tune Llama on my custom data, so how can I tokenize from a String and then detokenize the logits into a String?

I have looked at the Llama example for how to detokenize, but cannot find any clear documentation on how the implementation actually works for outputting results during training.

Thanks!

**Comment by LaurentMazare** [comment]
We use the `tokenizer` crate for this as you've probably noticed.
- For tokenizing, we use `encode`, see [here in the example](https://github.com/huggingface/candle/blob/5dbe46b389da4ba39131ce34752249cae640ad9e/candle-examples/examples/llama/main.rs#L192-L193). This results in a `Vec<u32>` that is then converted to a `Tensor`.
- For untokenizing, we use a hack so that we can do it in a streaming way rather than only being able to detokenize once the full text has been produced. The example does it [here](https://github.com/huggingface/candle/blob/5dbe46b389da4ba39131ce34752249cae640ad9e/candle-examples/examples/llama/main.rs#L234). Instead you probably want to use the [decode method](https://docs.rs/tokenizers/latest/tokenizers/tokenizer/struct.TokenizerImpl.html#method.decode).

Re documentation, I would suggest looking at the [tokenizer docs](https://docs.rs/tokenizers/latest/tokenizers/) as it's probably more relevant to this crate than to candle.

**Comment by EricLBuehler** [comment]
Ok, thank you! Regarding the tokenizer, what is the stop token ID? I see the start token is 1.

**Comment by LaurentMazare** [comment]
This depends on the tokenizer config, and the same goes for the start token. The best is probably not to assume anything here and to query for the token as is done in the [example that you referred to earlier](https://github.com/huggingface/candle/blob/5dbe46b389da4ba39131ce34752249cae640ad9e/candle-examples/examples/llama/main.rs#L190).

**Comment by EricLBuehler** [comment]
It seems like that would be </s> for that example. Thank you for your help! I will close the issue.
